src/libraries/System.Private.CoreLib/src/System/Guid.cs

```diff
@@ -769,33 +769,33 @@ public bool TryWriteBytes(Span<byte> destination)
         public override int GetHashCode()
         {
             // Simply XOR all the bits of the GUID 32 bits at a time.
-            return _a ^ Unsafe.Add(ref _a, 1) ^ Unsafe.Add(ref _a, 2) ^ Unsafe.Add(ref _a, 3);
+            ref int r = ref Unsafe.AsRef(in _a);
+            return r ^ Unsafe.Add(ref r, 1) ^ Unsafe.Add(ref r, 2) ^ Unsafe.Add(ref r, 3);
         }
 
         // Returns true if and only if the guid represented
         //  by o is the same as this instance.
         public override bool Equals(object? o)
         {
-            Guid g;
-            // Check that o is a Guid first
-            if (o == null || !(o is Guid))
-                return false;
-            else g = (Guid)o;
-
-            // Now compare each of the elements
-            return g._a == _a &&
-                Unsafe.Add(ref g._a, 1) == Unsafe.Add(ref _a, 1) &&
-                Unsafe.Add(ref g._a, 2) == Unsafe.Add(ref _a, 2) &&
-                Unsafe.Add(ref g._a, 3) == Unsafe.Add(ref _a, 3);
+            return o is Guid
+                && EqualsCore(this, Unsafe.Unbox<Guid>(o));
         }
 
-        public bool Equals(Guid g)
+        public bool Equals(Guid g) => EqualsCore(this, g);
+
+        private static bool EqualsCore(in Guid left, in Guid right)
         {
-            // Now compare each of the elements
-            return g._a == _a &&
-                Unsafe.Add(ref g._a, 1) == Unsafe.Add(ref _a, 1) &&
-                Unsafe.Add(ref g._a, 2) == Unsafe.Add(ref _a, 2) &&
-                Unsafe.Add(ref g._a, 3) == Unsafe.Add(ref _a, 3);
+            ref int rA = ref Unsafe.AsRef(in left._a);
+            ref int rB = ref Unsafe.AsRef(in right._a);
+
+            // Compare each element
+
+            if (rA != rB) { return false; }
+            if (Unsafe.Add(ref rA, 1) != Unsafe.Add(ref rB, 1)) { return false; }
```

Comment:
You can also change this to do the comparison as longs using `Unsafe.ReadUnaligned`. We had abandoned PR for that: https://github.com/dotnet/runtime/pull/35654

Reply:
Do you know why it was abandoned? I see the alignment discussion on that thread, but there doesn't seem to be any clear reason the PR was rejected.

Reply:
I guess Bruce had more important things to do than to finish it.

Reply:
Something like this I guess 🙂 
```csharp
private static bool EqualsCore_1(in Guid left, in Guid right)
{
    ref byte leftBytes = ref Unsafe.As<int, byte>(ref Unsafe.AsRef(in left._a));
    ref byte rightBytes = ref Unsafe.As<int, byte>(ref Unsafe.AsRef(in right._a));

    // early out if not equal
    if (Unsafe.ReadUnaligned<ulong>(ref leftBytes) != 
        Unsafe.ReadUnaligned<ulong>(ref rightBytes))
        return false;

    return Unsafe.ReadUnaligned<ulong>(ref Unsafe.Add(ref leftBytes, 8)) == 
           Unsafe.ReadUnaligned<ulong>(ref Unsafe.Add(ref rightBytes, 8));
}
```
a SIMD version of it slightly faster for equal guids but in most cases, I guess, we can "early out"

Reply:
Using 64-bit comparisons didn't seem to help matters - see https://github.com/dotnet/runtime/pull/44629#issuecomment-727076381.

Reply:
|              Equal_32Bit | 0.9002 ns | 0.0087 ns | 0.0082 ns |
|              Equal_64Bit | 1.1797 ns | 0.0109 ns | 0.0102 ns |

Interesting. https://github.com/dotnet/runtime/pull/35654 had this case as an improvement.